extensions/amp-story-dev-tools/0.1/amp-story-dev-tools-tab-preview.js

```diff
@@ -0,0 +1,751 @@
+
+    /** @private {!Element} container for the device previews */
+    this.devicesContainer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.storyUrl_ = this.element.getAttribute('data-story-url');
+    this.element.classList.add('i-amphtml-story-dev-tools-tab');
+
+    this.devicesContainer_ = htmlFor(
+      this.element
+    )`<div class="i-amphtml-story-dev-tools-devices-container"></div>`;
+    this.element.appendChild(this.devicesContainer_);
+
+    const chipListContainer = this.element.ownerDocument.createElement('div');
+    chipListContainer.classList.add(
+      'i-amphtml-story-dev-tools-device-chips-container'
+    );
+    this.element.appendChild(chipListContainer);
+    const chipList = this.element.ownerDocument.createElement('span');
+    chipList.classList.add('i-amphtml-story-dev-tools-device-chips');
+    chipListContainer.appendChild(chipList);
+    chipListContainer.appendChild(this.buildAddDeviceButton_());
+    chipListContainer.appendChild(this.buildHelpButton_());
+
+    parseDevices(
+      this.element.getAttribute('data-devices') || DEFAULT_DEVICES
+    ).forEach((device) => {
+      this.addDevice_(device.name);
+    });
+    this.repositionDevices_();
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.element.addEventListener('click', (e) => this.handleTap_(e.target));
+  }
+
+  /**
+   * Builds the add device button
+   * @private
+   * @return {!Element}
+   */
+  buildAddDeviceButton_() {
+    const addDeviceButton = buildDeviceChipTemplate(this.element);
+    addDeviceButton.classList.add('i-amphtml-story-dev-tools-add-device');
+    addDeviceButton.classList.add('i-amphtml-story-dev-tools-button');
+    addDeviceButton.setAttribute(
+      'data-action',
+      PREVIEW_ACTIONS.SHOW_ADD_DEVICE_DIALIG
+    );
+    addDeviceButton.querySelector(
+      '.i-amphtml-story-dev-tools-device-chip-name'
+    ).textContent = 'ADD DEVICE';
+    return addDeviceButton;
+  }
+
+  /**
+   * Builds the add device button
+   * @private
+   * @return {!Element}
+   */
+  buildHelpButton_() {
+    const addHelpButton = buildHelpButtonTemplate(this.element);
+    addHelpButton.setAttribute('data-action', PREVIEW_ACTIONS.SHOW_HELP_DIALOG);
+    return addHelpButton;
+  }
+
+  /**
+   * Creates a device layout for preview
+   * @private
+   * @param {!DeviceInfo} device
+   * @return {!Element}
+   */
+  buildDeviceLayout_(device) {
+    const deviceLayout = buildDeviceTemplate(this.element);
+    const devicePlayer = deviceLayout.querySelector('amp-story-player');
+    devicePlayer.setAttribute('width', device.width);
+    devicePlayer.setAttribute('height', device.height);
+    const storyA = devicePlayer.querySelector('a');
+    storyA.textContent = 'Story 1';
+    storyA.href = this.storyUrl_;
+    setStyles(devicePlayer, {
+      width: device.width + 'px',
+      height: device.height + 'px',
+    });
+    const playerImpl = new AmpStoryPlayer(this.win, devicePlayer);
+    playerImpl.load();
```

Comment:
It looks like the clipped frames are coming from this running during transition.
Setting a 300ms timeout helps but isn't a good solution. 
I wonder if we could async this by creating a pool of reused devices.
That might be too much work for this version but just an idea.

Reply:
Yes that's a really expensive operation! Lol we could make an extra device that we use when you add a device, but it will be complicating things much more. I will have to merge the code to sync the players navigation events later on and it would be a mess to also keep track of a hidden pool of devices... I don't like the idea of having "pools" everywhere, they always get hard to manage